### PR TITLE
Adding trace_attributes

### PR DESCRIPTION
**Pull request by kdiluca**
Current input can either be encoded_polyline or shape from a previous route request.

This returns edge info which include: edge names, id, base_data_id, weighted_upgrade, max_upward_grade and max_downward_grade in the trace_attributes response.

This is still under construction and will be undergoing a major refactor.

### CODE REVIEW

File: src/thor/trace_route_action.cc [hunk 1]
```diff
@@ -9,36 +9,228 @@ using namespace prime_server;
...
+    std::vector<baldr::PathLocation> through_loc;
+
+    // Form the trip path based on mode costing, origin, destination, and path edges
+    trip_path = thor::TripPathBuilder::Build(reader, mode_costing,
+                                                  path_infos,
+                                                  correlated.front(),
+                                                  correlated.back(),
+                                                  through_loc);
+
+    LOG_INFO(">>>>> route_match SUCCESS!!");
```
Comment (kevinkreiser) [inline src/thor/trace_route_action.cc]: remove

File: src/thor/trace_route_action.cc [hunk 2]
```diff
@@ -60,16 +252,16 @@ worker_t::result_t thor_worker_t::trace_route(
     results = (matcher->OfflineMatch(sequence));
   }
   LOG_INFO("Matched Results from input trace: ");
```
Comment (kevinkreiser) [inline src/thor/trace_route_action.cc]: remove

File: src/thor/trace_route_action.cc [hunk 3]
```diff
@@ -60,16 +252,16 @@ worker_t::result_t thor_worker_t::trace_route(
...
-  for(const auto& result : results) {
-    LOG_INFO(std::to_string(result.lnglat().first) + ", " + std::to_string(result.lnglat().second));
+  for (const auto& result : results) {
```
Comment (kevinkreiser) [inline src/thor/trace_route_action.cc]: remove

File: src/thor/trace_attributes_action.cc [hunk 4]
```diff
@@ -2,40 +2,121 @@
...
+            {"id", static_cast<uint64_t>(node.edge().id())},
+            {"names", names}
+          }));
+        }
+      }
+    }
+    auto json = json::map({
+      {"edges", edges}
+    });
+    //  if (id)
```
Comment (kevinkreiser) [inline src/thor/trace_attributes_action.cc]: dont you want to enable this? just requires passing the `const boost::optional<std::string>& id` from the caller

File: src/thor/trace_attributes_action.cc [hunk 5]
```diff
@@ -2,40 +2,121 @@
...
+  // If the exact points from a prior route that was run agains the Valhalla road network,
+  //then we can traverse the exact shape to form a path by using edge-walking algorithm
+  auto trip_path = route_match();
+  if (trip_path.node().size() == 0) {
+    //If no Valhalla route match, then use meili map matching
+    //to match to local route network. No shortcuts are used and detailed
+    //information at every intersection becomes available.
+    trip_path = map_match();
+  }
+  //result.messages.emplace_back(result);
```
Comment (kevinkreiser) [inline src/thor/trace_attributes_action.cc]: may as well delete this its never going on to the next stage from here

File: src/thor/trace_attributes_action.cc (hunk 4 above)
Comment (kevinkreiser) [inline src/thor/trace_attributes_action.cc]: or you could do it in the calling functino below

File: src/thor/expandfromnode.cc [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  const DirectedEdge* de = tile->directededge(node_info->edge_index());
+  for (uint32_t i = 0; i < node_info->edge_count(); i++, de++, edge_id++) {
+    // Skip shortcuts
+    if (de->is_shortcut()) {
+      continue;
+    }
+
+    // Process transition edge if previous edge was not from a transition
+    if (de->trans_down() || de->trans_up()) {
+      if (de->trans_down()) {
```
Comment (dgearhart) [inline src/thor/expandfromnode.cc]: looks like you did not have trace_actions branch merged in - these were removed

File: src/thor/trace_attributes_action.cc (hunk 4 above)
Comment (kdiluca) [inline src/thor/trace_attributes_action.cc]: Wasn't going to do that now.  Is that ok?

File: valhalla/thor/expandfromnode.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+                                const baldr::GraphTile* tile, const baldr::GraphId& node,
+                                const baldr::GraphId& stop_node,
+                                sif::EdgeLabel& prev_edge_label,
+                                float& elapsed_time,
+                                std::vector<PathInfo>& path_infos,
+                                const bool from_transition);
+ protected:
+
+  // Current travel mode
+  sif::TravelMode mode_;
```
Comment (kevinkreiser) [inline valhalla/thor/expandfromnode.h]: is this really needed, cant the above method just be static and we remove this member variable?

File: src/thor/trace_attributes_action.cc (hunk 5 above)
Comment (dgearhart) [inline src/thor/trace_attributes_action.cc]: if route match failed we need to do map match

File: src/thor/expandfromnode.cc (hunk 6 above)
Comment (kevinkreiser) [inline src/thor/expandfromnode.cc]: what are these empty `if`s for?

File: src/thor/trace_attributes_action.cc (hunk 5 above)
Comment (kdiluca) [inline src/thor/trace_attributes_action.cc]: Kevin means to delete the comment.

**Comment by kevinkreiser** [comment]
:rocket:

**Comment by dgearhart** [comment]
:shipit:

**Pull request merged**